node/src/client.rs

```diff
@@ -0,0 +1,396 @@
+//! MainRT Client abstractions
+
+#![allow(clippy::upper_case_acronyms)]
+
+pub use crate::service::{
+    FullBackend, FullClient, LightBackend, LightClient, MainExecutorDispatch,
+    StakingExecutorDispatch,
+};
+use primitives::{AccountId, Balance, Block, BlockNumber, Hash, Header, Index};
+use sc_client_api::{Backend as BackendT, BlockchainEvents, KeyIterator};
+use sp_api::{CallApiAt, NumberFor, ProvideRuntimeApi};
+use sp_blockchain::HeaderBackend;
+use sp_consensus::BlockStatus;
+use sp_runtime::{
+    generic::{BlockId, SignedBlock},
+    traits::{BlakeTwo256, Block as BlockT},
+    Justifications,
+};
+use sp_storage::{ChildInfo, PrefixedStorageKey, StorageData, StorageKey};
+use std::sync::Arc;
+
+/// A set of APIs that MainRT-like runtimes must implement.
+pub trait RuntimeApiCollection:
+    sp_transaction_pool::runtime_api::TaggedTransactionQueue<Block>
+    + sp_api::ApiExt<Block>
+    + sp_consensus_babe::BabeApi<Block>
+    + pallet_grandpa::fg_primitives::GrandpaApi<Block>
+    + sp_block_builder::BlockBuilder<Block>
+    + frame_system_rpc_runtime_api::AccountNonceApi<Block, AccountId, Index>
+    + pallet_transaction_payment_rpc_runtime_api::TransactionPaymentApi<Block, Balance>
+    + sp_api::Metadata<Block>
+    + sp_offchain::OffchainWorkerApi<Block>
+    + sp_session::SessionKeys<Block>
+    + sp_authority_discovery::AuthorityDiscoveryApi<Block>
+where
+    <Self as sp_api::ApiExt<Block>>::StateBackend: sp_api::StateBackend<BlakeTwo256>,
+{
+}
+
+impl<Api> RuntimeApiCollection for Api
+where
+    Api: sp_transaction_pool::runtime_api::TaggedTransactionQueue<Block>
+        + sp_api::ApiExt<Block>
+        + sp_consensus_babe::BabeApi<Block>
+        + pallet_grandpa::fg_primitives::GrandpaApi<Block>
+        + sp_block_builder::BlockBuilder<Block>
+        + frame_system_rpc_runtime_api::AccountNonceApi<Block, AccountId, Index>
+        + pallet_transaction_payment_rpc_runtime_api::TransactionPaymentApi<Block, Balance>
+        + sp_api::Metadata<Block>
+        + sp_offchain::OffchainWorkerApi<Block>
+        + sp_session::SessionKeys<Block>
+        + sp_authority_discovery::AuthorityDiscoveryApi<Block>,
+    <Self as sp_api::ApiExt<Block>>::StateBackend: sp_api::StateBackend<BlakeTwo256>,
+{
+}
```

Comment:
Looks like we have some redundancy here. I read it like: implement all the traits listed above for `Api` if `Api` is already implementing them!

Reply:
If I'm understanding correctly, this is a marker trait. As a suggestion, I think it makes sense to rename the generic parameter to something a bit clearer. `A` would be acceptable. If we're implementing generic runtime requirements, then maybe `Runtime` is appropriate?

Reply:
Good Question, :-). Right now client is adopted to upstream code.